### PR TITLE
Tickets/dm 31260

### PR DESCRIPTION
**Pull request by wvreeven**

### CODE REVIEW

File: python/lsst/ts/MTDome/encoding_tools.py [hunk 1]
```diff
@@ -21,33 +21,34 @@
...
+from .schema import (
+    AMCS_STATUS_JSCHEMA,
+    APSCS_STATUS_JSCHEMA,
+    COMMAND_JSCHEMA,
+    LCS_STATUS_JSCHEMA,
+    LWSCS_STATUS_JSCHEMA,
+    MONCS_STATUS_JSCHEMA,
+    RESPONSE_JSCHEMA,
+    THCS_STATUS_JSCHEMA,
+)
```
Comment (r-owen) [inline python/lsst/ts/MTDome/encoding_tools.py:36]: I am a bit unhappy about the repetition and the unclear connection between the schema constant names and the names by which you refer to the schema in your code. I'm not sure it's worth improving, but here are two ideas:

Make a schema registry (a dict of name: schema). This is a common and simple solution that scales well to adding new schemas. For instance add a new module `schemas/registry.py`:
```
__all__ = ["registry"]
import typing
registry: typing.Dict[str, str] = {}
```

Then each schema module is tweaked to look like this (notice that the names like AMCS_STATUS_JSHEMA are no longer needed, which is a good sign):
```
__all__ = []
import json
from .registry import registry
registry["AMCS"] = json.loads(
...
)
```
and `schemas/__init__.py` looks the same except you append `from .registry import *` to get the only public symbol from the schema sub-package. Note that it still must import all the schema modules _even though they have no public symbols_ in order to add the schemas to the registry.

A slightly fancier variant is to have a `register_schema` function that does some checking, such as rejecting duplicate names and perhaps even validating the schema, before adding the schema to the registry.

File: setup.py [hunk 2]
```diff
@@ -54,17 +53,8 @@
...
-        (os.path.join(data_files_path, "schema"), ["schema/apscs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/command.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/lcs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/lwscs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/moncs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/response.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/thcs_status.jschema"]),
-    ],
+    package_data={"": ["*.rst", "*.yaml", "*.xml"]},
+    data_files=[],
```
Comment (r-owen) [inline setup.py:57]: I suggest simply deleting `data_files=[],`

File: setup.py [hunk 3]
```diff
@@ -54,17 +53,8 @@
...
-        (os.path.join(data_files_path, "schema"), ["schema/amcs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/apscs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/command.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/lcs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/lwscs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/moncs_status.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/response.jschema"]),
-        (os.path.join(data_files_path, "schema"), ["schema/thcs_status.jschema"]),
-    ],
+    package_data={"": ["*.rst", "*.yaml", "*.xml"]},
```
Comment (r-owen) [inline setup.py:56]: Please try to figure out why flake8 didn't complain about unused variable `data_files_path`.

Once you get that figured out, please delete lines 41-43 (which define tools_path, base_prefix, and data_files_path).

**Review by r-owen — APPROVED** [review]
Thanks for moving the schemas into Python. That is a nice improvement to the deployment.

Please configure mypy to require type annotations. You have some functions that don't have any (e.g. in encoding_tools) and a lot that have no return type annotation. The command-line argument is `--disallow-untyped-defs`. You can make that a separate ticket if you prefer. It looks like I need to do that in salobj, as well.

I had a few other comments I hope you will consider.

File: tests/test_mock_controller.py [hunk 4]
```diff
@@ -172,14 +166,14 @@ async def verify_amcs_move(
         self.mock_ctrl.current_tai = self.mock_ctrl.current_tai + time_diff
         await self.write(command="statusAMCS", parameters={})
         self.data = await self.read()
-        amcs_status = self.data[LlcName.AMCS.value]
+        amcs_status = self.data[MTDome.llc_name.LlcName.AMCS.value]
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_mock_controller.py:169]: Good suggestions both. Done.

File: tests/test_elevation_motion.py [hunk 5]
```diff
@@ -54,7 +53,7 @@ async def prepare_elevation_motion(
         start_tai: `float`
             The current TAI time.
         """
-        self.elevation_motion = ElevationMotion(
+        self.elevation_motion = MTDome.mock_llc.mock_motion.ElevationMotion(
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_elevation_motion.py:56]: Done

File: tests/test_mock_controller.py [hunk 6]
```diff
@@ -588,7 +608,7 @@ async def prepare_all_llc(self):
 
         louver_id = 5
         target_position = 100
-        position = np.full(NUM_LOUVERS, -1.0, dtype=float)
+        position = np.full(MTDome.mock_llc.lcs.NUM_LOUVERS, -1.0, dtype=float)
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_mock_controller.py:611]: Done. I did not introduce a `mock_llc/constants.py`.

File: tests/test_mock_controller.py [hunk 7]
```diff
@@ -1098,18 +1118,18 @@ async def test_park(self):
...
+        amcs_status = self.data[MTDome.llc_name.LlcName.AMCS.value]
         self.assertEqual(
             amcs_status["status"]["status"],
-            LlcMotionState.PARKED.name,
+            MTDome.LlcMotionState.PARKED.name,
         )
         self.assertEqual(
             amcs_status["positionActual"],
-            PARK_POSITION,
+            MTDome.mock_llc.amcs.PARK_POSITION,
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_mock_controller.py:1128]: Done

File: tests/test_mock_controller.py [hunk 8]
```diff
@@ -1240,24 +1260,24 @@ async def test_status(self):
...
+        moncs_status = self.data[MTDome.llc_name.LlcName.MONCS.value]
         self.assertEqual(
             moncs_status["status"],
-            LlcMotionState.CLOSED.name,
+            MTDome.LlcMotionState.CLOSED.name,
         )
         self.assertEqual(
             moncs_status["data"],
-            [0.0] * NUM_MON_SENSORS,
+            [0.0] * MTDome.mock_llc.moncs.NUM_MON_SENSORS,
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_mock_controller.py:1270]: Done

File: tests/test_mock_controller.py [hunk 9]
```diff
@@ -1131,21 +1151,21 @@ async def test_setTemperature(self):
...
+        thcs_status = self.data[MTDome.llc_name.LlcName.THCS.value]
         self.assertEqual(
             thcs_status["status"],
-            LlcMotionState.OPEN.name,
+            MTDome.LlcMotionState.OPEN.name,
         )
         self.assertEqual(
             thcs_status["temperature"],
-            [temperature] * NUM_THERMO_SENSORS,
+            [temperature] * MTDome.mock_llc.thcs.NUM_THERMO_SENSORS,
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_mock_controller.py:1161]: Done

File: python/lsst/ts/MTDome/llc_configuration_limits/abstract_limits.py [hunk 10]
```diff
@@ -32,5 +32,5 @@ class AbstractLimits(ABC):
     """
 
     @abstractmethod
-    def validate(self, configuration_parameters):
+    def validate(self, configuration_parameters: dict):
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/MTDome/llc_configuration_limits/abstract_limits.py:35]: Done

File: python/lsst/ts/MTDome/llc_configuration_limits/common_amcs_and_lwscs_limits.py [hunk 11]
```diff
@@ -33,11 +33,15 @@ class CommonAmcsAndLwscsLimits(AbstractLimits):
     """
 
     @abstractmethod
-    def validate(self, configuration_parameters):
+    def validate(self, configuration_parameters: dict):
         pass
 
     # noinspection PyMethodMayBeStatic
-    def validate_common_parameters(self, configuration_parameters, common_limits):
+    def validate_common_parameters(
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/MTDome/llc_configuration_limits/common_amcs_and_lwscs_limits.py:40]: It is not necessary anymore and I removed the comment here and elsewhere too.

File: python/lsst/ts/MTDome/mock_controller.py [hunk 12]
```diff
@@ -170,7 +174,7 @@ async def stop(self):
         server.close()
         self.log.info("Done closing")
 
-    async def write(self, **data):
+    async def write(self, **data: Any):
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/MTDome/mock_controller.py:177]: It is not a dict, but a tuple of any type just like kwargs. I'll leave it as is for now.

File: python/lsst/ts/MTDome/mtdome_csc.py [hunk 13]
```diff
@@ -87,16 +89,18 @@ class MTDomeCsc(salobj.ConfigurableCsc):
 
     def __init__(
         self,
-        config_dir=None,
-        initial_state=salobj.State.STANDBY,
-        simulation_mode=0,
-        mock_port=None,
+        config_dir: str = None,
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/MTDome/mtdome_csc.py:92]: Done

File: setup.py (hunk 2 above)
Comment (wvreeven) [reply setup.py:57]: Done

File: setup.py (hunk 3 above)
Comment (wvreeven) [reply setup.py:56]: I honestly don't know why Mypy doesn't complain but I could remove those lines and two unused imports after that and Mypy still doesn't complain.

File: python/lsst/ts/MTDome/encoding_tools.py (hunk 1 above)
Comment (wvreeven) [reply python/lsst/ts/MTDome/encoding_tools.py:36]: It turns out that Mypy doesn't like the `__all__ = []` in the schema code files, so I removed those. Apart from that the simple solution works like a charm. Thanks very much.

Comment (r-owen) [reply python/lsst/ts/MTDome/encoding_tools.py]: Without an `__all__` every symbols is imported. I think all mypy is looking for is:
```
__all__: typing.List[str] = []
```

File: setup.py (hunk 3 above)
Comment (r-owen) [reply setup.py:51]: It's `flake8` I am worried about, not `mypy`. I fear `setup.cfg` may needlessly be excluding `setup.py` from flake8 checking.

**Pull request merged**